File: codes/statisticalAnalysisLearning/statisticalLearning/FirstCourseMachineLearningCode/R/chapter5/bayesclass.R
## bayesclass.R
# From A First Course in Machine Learning, Chapter 5.
# Francois-Xavier Briol, 04/06/16 [f-x.briol@warwick.ac.uk]
# Bayesian classifier
rm(list=ls(all=TRUE))

## Load the data
setwd("~/data/bc_data") ## Might need to change the path here
X = as.matrix(read.csv(file="X.csv",header=FALSE))
t = as.matrix(read.csv(file="t.csv",header=FALSE))

# Plot the data
cl = as.vector(unique(t))
col = c('red','green','blue')
pos = which(t==cl[1])
plot(X[pos,1],X[pos,2],xlim=c(-3,7),ylim=c(-6,6),pch=16,col=col[1])
for(c in 2:length(cl)){
  pos = which(t==cl[c])
  points(X[pos,1],X[pos,2],col=col[c],pch=16)
}

## Fit class-conditional Gaussians for each class
# Using the Naive (independence) assumption
class_mean = matrix(NA,ncol=dim(X)[2],nrow=length(cl))
class_var = matrix(NA,ncol=dim(X)[2],nrow=length(cl))
for(c in 1:length(cl)){
  pos = which(t==cl[c])
  # Find the means
  class_mean[c,] = apply(X=X[pos,],MARGIN=2,FUN=mean)
  class_var[c,] = apply(X=X[pos,],MARGIN=2,FUN=var)
}

## Compute the predictive probabilities
gridvalsX = seq(-6,6,0.1)
gridvalsY = seq(-6,6,0.1)
mesh = meshgrid(gridvalsX,gridvalsY)
Xv = mesh$X; Yv = mesh$Y

Probs = list()
sumProbs = matrix(0,ncol=dim(Xv)[[2]],nrow=dim(Xv)[[1]])

for(c in 1:length(cl)){
  temp = cbind(as.vector(Xv)-class_mean[c,1], as.vector(Yv)-class_mean[c,2])
  tempc = diag(class_var[c,])
  const = -log(2*pi) - log(det(tempc))
  Probs[[c]] = matrix(exp(const - (0.5*diag(temp%*%solve(tempc)%*%t(temp)))),nrow=dim(Xv)[[1]],ncol=dim(Xv)[[2]],byrow=TRUE)
  sumProbs = sumProbs+Probs[[c]]
}
for(c in 1:length(cl)){
  Probs[[c]] = Probs[[c]]/sumProbs
}

## Plot the predictive contours
par(mfrow=c(1,3))
for(i in 1:length(cl)){
  cl = as.vector(unique(t))
  col = c('red','green','blue')
  pos = which(t==cl[1])
  plot(X[pos,1],X[pos,2],xlim=c(-3,7),ylim=c(-6,6),pch=16,
       col=col[1],main=paste("Prob. Cont. Class",i))
  for(c in 2:length(cl)){
    pos = which(t==cl[c])
    points(X[pos,1],X[pos,2],col=col[c],pch=16)
  }
  contour(gridvalsX,gridvalsY,Probs[[i]],add=TRUE)
}
par(mfrow=c(1,1))


## Repeat without Naive assumption
class_var = list()
for(c in 1:length(cl)){
  pos = which(t==cl[c])
  # Find the means
  class_mean[c,] = apply(X=X[pos,],MARGIN=2,FUN=mean)
  class_var[[c]] = cov(X[pos,])
}

## Compute the predictive probabilities
gridvalsX = seq(-6,6,0.1)
gridvalsY = seq(-6,6,0.1)
mesh = meshgrid(gridvalsX,gridvalsY)
Xv = mesh$X; Yv = mesh$Y

Probs = list()
sumProbs = matrix(0,ncol=dim(Xv)[[2]],nrow=dim(Xv)[[1]])

for(c in 1:length(cl)){
  temp = cbind(as.vector(Xv)-class_mean[c,1], as.vector(Yv)-class_mean[c,2])
  tempc = class_var[[c]]
  const = -log(2*pi) - log(det(tempc))
  Probs[[c]] = matrix(exp(const - (0.5*diag(temp%*%solve(tempc)%*%t(temp)))),nrow=dim(Xv)[[1]],ncol=dim(Xv)[[2]],byrow=TRUE)
  sumProbs = sumProbs+Probs[[c]]
}
for(c in 1:length(cl)){
  Probs[[c]] = Probs[[c]]/sumProbs
}

## Plot the predictive contours
par(mfrow=c(1,3))
for(i in 1:length(cl)){
  cl = as.vector(unique(t))
  col = c('red','green','blue')
  pos = which(t==cl[1])
  plot(X[pos,1],X[pos,2],xlim=c(-3,7),ylim=c(-6,6),pch=16,
       col=col[1],main=paste("Prob. Cont. Class",i))
  for(c in 2:length(cl)){
    pos = which(t==cl[c])
    points(X[pos,1],X[pos,2],col=col[c],pch=16)
  }
  contour(gridvalsX,gridvalsY,Probs[[i]],add=TRUE)
}
par(mfrow=c(1,1))

File: codes/statisticalAnalysisLearning/statisticalLearning/FirstCourseMachineLearningCode/matlab/chapter5/bayesclass.m
%% bayesclass.m
% From A First Course in Machine Learning, Chapter 5.
% Simon Rogers, 01/11/11 [simon.rogers@glasgow.ac.uk]
% Bayesian classifier
clear all;close all;

%% Load the data
load ../data/bc_data

% Plot the data

cl = unique(t);
col = {'ko','kd','ks'}
fcol = {[1 0 0],[0 1 0],[0 0 1]};
figure(1);
hold off
for c = 1:length(cl)
    pos = find(t==cl(c));
    plot(X(pos,1),X(pos,2),col{c},...
        'markersize',10,'linewidth',2,...
        'markerfacecolor',fcol{c});
    hold on
end
xlim([-3 7])
ylim([-6 6])



%% Fit class-conditional Gaussians for each class
% Using the Naive (independence) assumption
for c = 1:length(cl)
    pos = find(t==cl(c));
    % Find the means
    class_mean(c,:) = mean(X(pos,:));
    class_var(c,:) = var(X(pos,:),1);
end


%% Compute the predictive probabilities
[Xv,Yv] = meshgrid(-3:0.1:7,-6:0.1:6);
Probs = [];
for c = 1:length(cl)
    temp = [Xv(:)-class_mean(c,1) Yv(:)-class_mean(c,2)];
    tempc = diag(class_var(c,:));
    const = -log(2*pi) - log(det(tempc));
    Probs(:,:,c) = reshape(exp(const - 0.5*diag(temp*inv(tempc)*temp')),size(Xv));;
end

Probs = Probs./repmat(sum(Probs,3),[1,1,3]);

%% Plot the predictive contours
figure(1);hold off
for i = 1:3
    subplot(1,3,i);
    hold off
    for c = 1:length(cl)
        pos = find(t==cl(c));
        plot(X(pos,1),X(pos,2),col{c},...
            'markersize',10,'linewidth',2,...
            'markerfacecolor',fcol{c});
        hold on
    end
    xlim([-3 7])
    ylim([-6 6])
    
    contour(Xv,Yv,Probs(:,:,i));
    ti = sprintf('Probability contours for class %g',i);
    title(ti);
end


%% Repeat without Naive assumption
class_var = [];
for c = 1:length(cl)
    pos = find(t==cl(c));
    % Find the means
    class_mean(c,:) = mean(X(pos,:));
    class_var(:,:,c) = cov(X(pos,:),1);
end


%% Compute the predictive probabilities
[Xv,Yv] = meshgrid(-3:0.1:7,-6:0.1:6);
Probs = [];
for c = 1:length(cl)
    temp = [Xv(:)-class_mean(c,1) Yv(:)-class_mean(c,2)];
    tempc = class_var(:,:,c);
    const = -log(2*pi) - log(det(tempc));
    Probs(:,:,c) = reshape(exp(const - 0.5*diag(temp*inv(tempc)*temp')),size(Xv));;
end

Probs = Probs./repmat(sum(Probs,3),[1,1,3]);

%% Plot the predictive contours
figure(1);hold off
for i = 1:3
    subplot(1,3,i);
    hold off
    for c = 1:length(cl)
        pos = find(t==cl(c));
        plot(X(pos,1),X(pos,2),col{c},...
            'markersize',10,'linewidth',2,...
            'markerfacecolor',fcol{c});
        hold on
    end
    xlim([-3 7])
    ylim([-6 6])
    
    contour(Xv,Yv,Probs(:,:,i));
    ti = sprintf('Probability contours for class %g',i);
    title(ti);
end